File: c/validate_credit_card.c
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#include "input.h"

#define LENGTH 16

char *get_input(void);
bool luhn_algorithm(char *);

int main(void)
{
    char *card_numbers = get_input();

    if (!(luhn_algorithm(card_numbers)))
    {
        printf("INVALID\n");
        return 1;
    }

    printf("VALID\n");
    return 0;
}

char *get_input(void)
{
    char *input = NULL;
    int count = 0;

    do
    {
        input = get_string("Number (16 digits): ");

        for (int i = 0; i < LENGTH && isdigit(input[i]); i++)
        {
            count = i;
        }
    }
    while (strlen(input) != count && count != LENGTH);

    return input;
}

bool luhn_algorithm(char *card)
{
    int addition = 0; 

    for (int i = 0, product = 0; i < LENGTH; i += 2)
    {
        if ((product = (card[i] - '0') * 2) > 9)
        {
            product -= 9;
        }

        addition += product + card[i + 1] - '0';
    }

    return (!(addition % 10));
}

File: python/validate_credit_card.py
from typing import Final, List
from sys import exit

LENGTH: Final[int] = 16


def main() -> None:
    card_numbers: List[int] = get_input()

    if not luhn_algorithm(card_numbers):
        print('INVALID')
        exit(1)

    print('VALID')
    exit(0)


def get_input() -> int:
    while True:
        try:
            numbers: List[int] = [
                int(number) for number in input('Number (16 digits): ')
            ]
        except ValueError:
            print('Not an integer.')
        else:
            if numbers and len(numbers) == LENGTH:
                return numbers


def luhn_algorithm(numbers: List[int]) -> bool:
    addition: int = 0
    product: int = 0

    for i in range(0, LENGTH, 2):
        if (product := numbers[i] * 2) > 9:
            product -= 9

        addition += product + numbers[i + 1]

    return not addition % 10


if __name__ == '__main__':
    main()
